src/main/java/com/redhat/pantheon/servlet/GetObjectByUUID.java

```diff
@@ -39,41 +44,111 @@
         } catch (ItemNotFoundException infex) {
             response.sendError(HttpServletResponse.SC_NOT_FOUND);
             return;
         } catch (RepositoryException e) {
             throw new ServletException(e);
         }
     }
+
+    /**
+     * Accepts incoming URL parameter "dereference" as a String and builds a Map of properties to dereference, based
+     * off of the sling resource type that the property is found on. For example:
+     * Input: "pantheon/moduleLocale:draft,pantheon/moduleLocale:released,pantheon/moduleVersion:product"
+     * Output:
+     *  {
+     *      "pantheon/moduleLocale": ["draft","released"],
+     *      "pantheon/moduleVersion": ["product"]
+     *  }
+     * @param dereference The incoming dereference specification
+     * @return A Map of Sets containing the sling resource type as the Key, and the Set of properties to dereference
+     *  for that resource type as the Value.
+     */
+    private Map<String, Set<String>> buildDereferenceMap(String dereference) {
+        Map<String, Set<String>> map = new HashMap<>();
+        for (String token : dereference.split(",")) {
+            String[] pair = token.split(":", 2);
+            Set<String> props = map.get(pair[0]);
+            if (props == null) {
+                props = new HashSet<>();
+                map.put(pair[0], props);
+            }
+            props.add(pair[1]);
+        }
+        return map;
+    }
+
+    private Resource getResourceByUuid(String uuid) throws RepositoryException {
+        Node foundNode = request.getResourceResolver()
+                .adaptTo(Session.class)
+                .getNodeByIdentifier(uuid);
+
+        // turn the node back into a resource
+        Resource foundResource = request.getResourceResolver()
+                .getResource(foundNode.getPath());
+
+        return foundResource;
+    }
+
+    private Map<String, Object> resourceToMapRecursive(Resource resource, Long depth, Map<String, Set<String>> dereferenceMap) throws RepositoryException {
+        Map<String, Object> ret = new HashMap<>();
+        ret.putAll(resource.getValueMap());
+
+        if (depth > 0) {
+            for (Resource child : resource.getChildren()) {
+                ret.put(child.getName(), resourceToMapRecursive(child, depth - 1, dereferenceMap));
+            }
+
+            String type = resource.getResourceType();
+            Set<String> referenceFields = dereferenceMap.get(type);
+            if (referenceFields != null) {
+                for (Map.Entry<String, Object> entry : resource.getValueMap().entrySet()) {
+                    if (referenceFields.contains(entry.getKey())) {
+                        Resource child = getResourceByUuid((String) entry.getValue());
+                        ret.put(entry.getKey(), resourceToMapRecursive(child, depth - 1, dereferenceMap));
+                    }
+                }
+            }
+        }
+
+        ret.remove("jcr:data");
```

Comment:
Why are you removing this specific property?

Reply:
I don't remember the specific issue, but when I was running queries against nodes with binary data (an image perhaps?), there was "something" preventing me from viewing the result of the query. I don't remember if the query was failing to return anything because the data couldn't be represented in JSON, or if it was the browser that was unable to render a human-readable result, but that was the sort of issue that precipitated this. My thought was that generally, we will not need the `jcr:data` to be returned with this query - and if we do, it probably ought to be its own API call (perhaps in the form of a "download source file" dialog that we don't have on the radar currently).

Reply:
OK, just be aware we'll hit the same problem with other binary fields with a different name. If you are going to leave this in, I'd suggest a long comment to remind us of the reasoning.